WIP: singularity workarounds

octomike opened the pull request:
We are using singularity and the docker container needs some minor changes to be usable:

1. At least for the longitudinal batch at runtime the standalone code will replace the content of the pseudo compiled ".m" file with the content of a ".txt" file with the same name. Since with singularity the content of the container is read-only, this will fail and we replace at build-time. In theory this should also prevent docker runs from creating tiny unnamed volumes.

2. The container content often belongs to `root` with singularity and the CAT12 mex folder needs executable bit set for groups and others.

I also added a minimal example for a singularity invocation in the README.

WIP, we still haven't verified a batch completely, but it's looking good so far.

octomike committed:
Replace "compiled" m-file to work with read-only containers

octomike committed:
ENH: chmod +x for mex files when owned by root (singularity)

octomike committed:
ENH: add singularity example in README

jhuguetn commented on Dockerfile:
```diff
@@ -38,7 +38,9 @@ RUN wget --progress=bar:force -P /tmp http://www.neuro.uni-jena.de/cat12/${CAT_F
  && mv /opt/${CAT_FULLVERSION}_${MATLAB_VERSION}_MCR_Linux /opt/spm \
  && /opt/spm/run_spm12.sh ${MCRROOT} --version \
  && chmod +x /opt/spm/spm12 /opt/spm/*.sh \
+ && chmod +x /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/CAT.glnx86/CAT_* \
```
Funny, those files were executable for all users in older CAT versions

octomike replied:
They are in the git repo, but it's missing the permission bits after the build.

The call to `/opt/spm/run_spm12.sh ${MCRROOT} --version ` is what's unpacking `spm12.ctf` implicitly, right?

`zipinfo` lists the files as non-executable in the current standalone zip.

```
root@981c119aa9aa:~/CAT12.8.1_r2042_R2017b_MCR_Linux# zipinfo spm12.ctf  | grep -e w32/CAT_Surf2Vol -e lnx86/CAT_Surf2Vol
-rw----     0.0 fat  1993150 b- defN 22-Jul-19 00:18 fsroot/home/gaser/gaser/spm/spm12/toolbox/cat12/CAT.glnx86/CAT_Surf2Vol
-rwx---     0.0 fat  1759485 b- defN 22-Jul-19 00:18 fsroot/home/gaser/gaser/spm/spm12/toolbox/cat12/CAT.w32/CAT_Surf2Vol.exe
```

jhuguetn commented on Dockerfile:
```diff
@@ -38,7 +38,9 @@ RUN wget --progress=bar:force -P /tmp http://www.neuro.uni-jena.de/cat12/${CAT_F
  && chmod +x /opt/spm/spm12 /opt/spm/*.sh \
+ && chmod +x /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/CAT.glnx86/CAT_* \
  && rm -rf /tmp/*
+RUN cp /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.txt /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.m
```
I don't like permanently replacing a compiled Matlab file by an existing script file (text). Would it be possible to rename it to something else such as `cat_long_main2.m`?

octomike replied:
If we run the longitudinal segmentation on your vanilla image with singularity, it immediately fails with:

```
------------------------------------------------------------------------
18-Jan-2023 18:29:15 - Running job #1
------------------------------------------------------------------------
18-Jan-2023 18:29:15 - Running 'CAT12: Segment longitudinal data'
cp: cannot remove '/opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.m': Read-only file system
If you do not have write permissions, the administrator should copy the /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.txt file to /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.m after installing the precompiled version. This prevents overwriting the read-only file.
```

Until now I thought this was part of the `cat_standalone.sh` script, but I'm wrong. At least I cannot find an explicit `cp`. So it must be either spm trying to do this (move to .m if .txt exists?) or some dark magic from matlab / the mcr itself.

octomike replied:
Aaah.. they removed the accompanying `.txt` file here https://github.com/ChristianGaser/cat12/commit/febd62595be99f2bad049d9df93e920903fb3b06

So that line can be removed then as it shouldn't be necessary anymore for future builds. Or it could be changed to just deleting the `.txt` file if someone decides to use your Dockerfile to build a historic version (quite likely actually).

jhuguetn replied:
As is not an issue anymore (in current/future CAT releases) I would prefer having the line removed and keep a leaner (readable) Dockerfile script. Thanks @octomike

jhuguetn replied:
I tried the newer release of CAT (12.8.2) and this problem remains an open issue with Singularity. I would like to better understand what is going on under the curtains, why are both -a matlab script `cat_long_main.m` renamed to .txt and a compiled version of the same script- needed and swapped (replaced) when longitudinal segmentation is executed.

octomike committed:
ENH: remove sidecar txt file for cat_long_main.txt

This has been fixed in upstream. Historical builds might still
encounter this to be necessary for using a read-only container
system such as Singularity.

octomike commented:
> WIP, we still haven't verified a batch completely, but it's looking good so far.

Update: With the latest changes in https://github.com/octomike/matlab-compiler-runtime/ we were able to run a regular segmentation and a longitudinal segmentation with this stripped down "core" version. Still, it's a quite hack, so further testing should be done, before this get merged.

jhuguetn commented on README.md:
````diff
@@ -29,6 +29,21 @@ Example (segmentation of T1w image)
+mode, because the MCR will write into that folder.
+
+```bash
+singularity build cat12-latest.sif docker://jhuguetn/cat12:latest
+singularity run --cleanenv --contain \
+  -B $PWD:/data \
+  -B $HOME/.matlab \
+  cat12-latest.sif \
+  -s /opt/spm -m /opt/mcr/v93 -b /opt/spm/standalone/cat_standalone_segment_long.m /data/my_dataset/sub-01/ses-0{1,2,3}/anat/sub-01_T1w.nii
````
You don't have to specify SPM and MCR root directories, they are set as environment variables. We can keep a simpler example,

```suggestion
singularity build cat12-latest.sif docker://jhuguetn/cat12:latest
singularity run --cleanenv --contain \
  -B $PWD:/data \
  -B $HOME/.matlab \
  cat12-latest.sif \
  -b /opt/spm/standalone/cat_standalone_segment_long.m /data/my_dataset/sub-01/ses-0{1,2,3}/anat/sub-01_T1w.nii
```

jhuguetn commented on Dockerfile:
```diff
@@ -38,7 +38,9 @@ RUN wget --progress=bar:force -P /tmp http://www.neuro.uni-jena.de/cat12/${CAT_F
  && chmod +x /opt/spm/spm12 /opt/spm/*.sh \
+ && chmod +x /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/CAT.glnx86/CAT_* \
  && rm -rf /tmp/*
+RUN rm -f /opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_main.txt
```
This workaround does not work when run with Docker (latest CAT12 release),
```------------------------------------------------------------------------
07-Feb-2023 15:02:40 - Running job #1
------------------------------------------------------------------------
07-Feb-2023 15:02:41 - Running 'CAT12: Segment longitudinal data'
07-Feb-2023 15:02:41 - Failed  'CAT12: Segment longitudinal data'
Error using fread
Invalid file identifier. Use fopen to generate a valid file identifier.
In file "/opt/spm/spm12_mcr/home/gaser/gaser/spm/spm12/toolbox/cat12/cat_long_multi_run.m" (???), function "cat_long_multi_run" at line 187.

The following modules did not run:
Failed: CAT12: Segment longitudinal data

Execution failed: /tmp/cat_1.m
Bye for now...

Error using MATLABbatch system
Job execution failed. The full log of this run can be found in MATLAB command window, starting with the lines (look for the line showing the exact #job as displayed in                              this error message)
------------------
Running job #1
------------------

Error in MATLABbatch system
```

jhuguetn commented:
Proposal addressed at https://github.com/jhuguetn/cat12-docker/pull/8, closing this PR. Thanks a lot @octomike for your contribution.
